agent/node.go

```diff
@@ -600,7 +601,26 @@ func (n *Node) runManager(ctx context.Context, securityConfig *ca.SecurityConfig
 			ElectionTick:   n.config.ElectionTick,
 		})
 		if err != nil {
-			return err
+			logrus.Errorf("failed to starting manager: %v", err)
+			// Something went wrong, but it might be fixed later on
+			// by user based on the error log. So rather than quit,
+			// we keep waiting for my role to be changed back if
+			// promotion failed. In the meantime, the agent part still
+			// work. The only exception is that the node is being started as
+			// manager role and caller is waiting for manager part ready.
+
+			// TODO(jinuxstyle): change role back to agent if promotion
+			// failed in a self-healing way.
+
+			if initialRole == ca.ManagerRole && ready != nil {
+				return err
+			}
+
+			n.waitRole(ctx, ca.AgentRole)
+			if ctx.Err() != nil {
```

Comment:
I agree with the approach you suggested. Indeed, the waitRole() used to return error until commit 294ade7bd976796. I don't if there is any strong reason for that. Ping @LK4D4 for comment.